working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,166 @@
+
+Since a mixin application class cannot contain static members, there is no risk of the introducing a conflict between a constructor name and a static member. 
+
+### Goal
+
+We combine the forwarding constructor feature with default constructors and automatic initialization of declared instance variables. This *unifies* default constructors and mixin application forwarding constructors *and* adds more power to the default constructor for simple classes.
+
+We expect this to happen after non-nullable types have been added to the language, so the proposal is written in terms of that, with a few guidelines for how legacy libraries work where nullability matters.
+
+## Proposal
+
+Let *C* be a class declaration with name `C` and superclass *S* with name `S` and which *declares* no constructor.
+
+It is a compile-time error if:
+
+- *C* declares a private-named and potentially non-nullable instance variable with no initializer expressions. 
+- Or in **NNBD-legacy mode**,  *C* declares a private-named and final instance variable with no initializer expression.
+
+Otherwise, for each constructor *s* of the superclass, named `S` or `S.id`, satisfying all of the following:
+
+1. If *s* is a named constructor with name `S.id` then:
+   1.  *s* is accessible ( `id` is not a private identifier or *S* is not declared in a different library), and
+   2.  *C* does not declare a static member named *id*.
+2. The *s* constructor is a generative constructor.
+3. *C* does not declare an instance variable with no initializer expression and with the same basename as a required named parameter of *s*.
+4. Either *C* does not declare any instance variables or *s* does not have any optional positional parameters (required because we can't declare a constructor with both optional position parameters and named parameters).
+
+add a generative default constructor *c* to *C* with the corresponding name `C` or `C.id`, the following parameters:
+
+- For each non-private instance variable with no initializer expression declared by *C*, where `v` is the name of the instance variable, the constructor has a named initializing formal parameter `this.v` with no default value. The named parameter is *required* if the instance variable has a potentially non-nullable type.
+- For each named parameter of *s*, *c* has a corresponding parameter with the same name, type, optionality and default value (if optional), unless a named initializing formal parameter with the same name was declared above (in which case the parameter of *s* must be optional per requirement 4. above).
```

Comment:
This makes it a breaking change to add an optional named parameter to a constructor.

Reply:
It makes it a breaking change to add a *required* parameter to the superclass constructor, but that was always the case.

Adding a new field named `x` to the subclass may change the meaning of calling `Foo(x: 42)`, if it previously referred to a superclass constructor parameter. 
That's also not new, you change the meaning of an existing constructor. You should be aware of the super-constructor parameters. The only issue is that the change is not directly visible in the source, so you might not be aware.

That kind of conflict is unavoidable if we want to forward *any* parameter to the superclass constructor, which I do believe is valuable.